### DISCUSSION TITLE
External Dependencies at Build Time

### DISCUSSION THREAD

**Discussion by mys721tx** [Homebrew/discussions#4319] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.6-8-g0a6dffd
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0a6dffdcf96588ad064b7c0395458821bf27cf88
Last commit: 65 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c26f1c8714c231ee6796915f033ba229c89af3d6
Core tap last commit: 4 weeks ago
Core tap branch: master
Core tap JSON: 11 Mar 16:50 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.whehDGlItY/org.xquartz:0
HOMEBREW_EDITOR: vim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  blast
  python@3.11

Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
  ?? Formula/picotool.rb
```

### Description of issue

I am writing a formula for [raspberrypi/picotool](https://github.com/raspberrypi/picotool) and it depends on [raspberrypi/pico-sdk](https://github.com/raspberrypi/pico-sdk) at build time. Since the SDK is for RP2040, we probably should not make a formula for it. What is the standard way to handle an external dependency at build time?

**Comment by ZhongRuoyu** [comment, marked as answer]
You may make `pico-sdk` a resource. See [`telnet.rb`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/telnet.rb) for an example, where `libtelnet` is added as a resource and built in the `install` method below.

**Reply by mys721tx** [reply]
Thanks! Homebrew/homebrew-core#125398 submitted.

**Comment by AltonMusinski3** [comment]
If you have a build-time dependency that cannot be installed through Homebrew, you can add the dependency as a resource to your formula. A resource is a file or a directory that is downloaded or copied into the build directory before building the formula.

In your case, you can add the Pico SDK as a resource to your formula, so that it is available during the build process. Here is an example of how you can do it:
class MyFormula < Formula
  desc "My formula"
  homepage "https://example.com"
  url "https://example.com/archive/v1.0.tar.gz"
  sha256 "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"

  resource "pico-sdk" do
    url "https://github.com/raspberrypi/pico-sdk/archive/master.tar.gz"
    sha256 "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
  end

  def install
    # Copy the Pico SDK into the build directory
    resource("pico-sdk").stage do
      cp_r ".", buildpath/"pico-sdk"
    end

    # Build your formula using the Pico SDK
    system "make", "SDK_PATH=#{buildpath}/pico-sdk", "install"
  end

  test do
    # ...
  end
end

In this example, the Pico SDK is added as a resource to the formula using the resource method. The stage block is used to download and extract the SDK into the build directory. The install method is then modified to pass the path to the SDK as a variable to the build command.

You can replace the URL and SHA256 checksum of the Pico SDK with the ones that are appropriate for your use case.
